server/src/main/java/org/apache/druid/server/coordination/SegmentLoadDropHandler.java

```diff
@@ -100,13 +119,20 @@ public SegmentLoadDropHandler(
       SegmentLoaderConfig config,
       DataSegmentAnnouncer announcer,
       SegmentManager segmentManager,
-      ScheduledExecutorService exec
+      ThreadPoolExecutor standardExec,
+      ThreadPoolExecutor turboExec,
+      ScheduledExecutorService scheduledExecutorService
   )
   {
     this.config = config;
     this.announcer = announcer;
     this.segmentManager = segmentManager;
-    this.exec = exec;
+    this.standardExec = standardExec;
+    this.turboExec = turboExec;
+    this.scheduledExecutorService = scheduledExecutorService;
+
+    this.standardExec.allowCoreThreadTimeOut(true);
+    this.turboExec.allowCoreThreadTimeOut(true);
```

Comment:
We probably don't need this if we do the following.

Since non-core threads can always be terminated, how about we just use a single thread pool with
`coreSize = numLoadingThreads` and `maxSize = max(numLoadingThreads, numBootstrapThreads)`.

We would submit more requests only when a server is in turbo mode.
In normal mode, server would automatically terminate threads and use a smaller thread pool size.

Reply:
I think this conflicts with the other comment to use a `ScheduledThreadPoolExecutor` instead of the standardExec, as scheduled executors do not respect any `maximumPoolSize`, based on their javadoc. We would need at least two pools due to this (a scheduled normal pool and turbo pool).

> We would submit more requests only when a server is in turbo mode.
I also did not understand this, if we have `maxSize = max(numLoadingThreads, numBootstrapThreads)` and we are currently in normal mode, if an new request comes, would we not add a task for it? Or do you mean that the coordinator would only send the correct number of requests so it doesn't matter?

Reply:
Hmm, I see. In that case, let's just use a `ThreadPoolExecutor loadExec` and a `ScheduledExecutorService dropExec`. The `dropExec` can have a fixed small number of threads, say 2 or 5, since drops are fast.